File: client/src/test/scala/stasis/test/specs/unit/client/tracking/trackers/DefaultRecoveryTrackerSpec.scala
package stasis.test.specs.unit.client.tracking.trackers

import akka.actor.typed.scaladsl.Behaviors
import akka.actor.typed.{ActorSystem, Behavior, SpawnProtocol}
import akka.stream.scaladsl.Sink
import akka.util.Timeout
import org.scalatest.{Assertion, BeforeAndAfterAll}
import org.scalatest.concurrent.Eventually
import stasis.client.model.TargetEntity
import stasis.client.tracking.state.RecoveryState
import stasis.client.tracking.state.RecoveryState.ProcessedTargetEntity
import stasis.client.tracking.trackers.DefaultRecoveryTracker
import stasis.core.persistence.backends.memory.EventLogMemoryBackend
import stasis.shared.ops.Operation
import stasis.test.specs.unit.AsyncUnitSpec
import stasis.test.specs.unit.client.Fixtures
import stasis.test.specs.unit.core.telemetry.MockTelemetryContext

import scala.concurrent.duration._

class DefaultRecoveryTrackerSpec extends AsyncUnitSpec with Eventually with BeforeAndAfterAll {
  "A DefaultRecoveryTracker" should "track recovery events" in withRetry {
    val tracker = createTracker()

    implicit val operation: Operation.Id = Operation.generateId()

    val entity = Fixtures.Metadata.FileOneMetadata.path

    val targetEntity = TargetEntity(
      path = entity,
      existingMetadata = Fixtures.Metadata.FileOneMetadata,
      currentMetadata = None,
      destination = TargetEntity.Destination.Default
    )

    val initialState = tracker.state.await
    initialState should be(empty)

    tracker.entityExamined(entity = entity, metadataChanged = false, contentChanged = false)
    tracker.entityCollected(entity = targetEntity)
    tracker.entityProcessingStarted(entity = entity, expectedParts = 1)
    tracker.entityPartProcessed(entity = entity)
    tracker.entityProcessed(entity = entity)
    tracker.metadataApplied(entity = entity)
    tracker.failureEncountered(new RuntimeException("Test failure #1"))
    tracker.failureEncountered(entity = Fixtures.Metadata.FileTwoMetadata.path, new RuntimeException("Test failure #2"))
    tracker.completed()

    eventually[Assertion] {
      val state = tracker.state.await.get(operation) match {
        case Some(value) => value
        case None        => fail(s"Expected state for operation [$operation] but none was found")
      }

      state.operation should be(operation)

      state.entities should be(
        RecoveryState.Entities(
          examined = Set(entity),
          collected = Map(entity -> targetEntity),
          pending = Map.empty,
          processed = Map(
            entity -> ProcessedTargetEntity(expectedParts = 1, processedParts = 1)
          ),
          metadataApplied = Set(entity),
          failed = Map(
            Fixtures.Metadata.FileTwoMetadata.path -> "RuntimeException - Test failure #2"
          )
        )
      )

      state.failures should be(Seq("RuntimeException - Test failure #1"))
      state.completed should not be empty
    }
  }

  it should "provide state updates" in withRetry {
    val tracker = createTracker()

    implicit val operation: Operation.Id = Operation.generateId()

    val entity = Fixtures.Metadata.FileOneMetadata.path

    val initialState = tracker.state.await
    initialState should be(empty)

    val expectedUpdates = 3
    val updates = tracker.updates(operation).take(expectedUpdates.toLong).runWith(Sink.seq)
    await(100.millis, withSystem = system)

    tracker.entityExamined(entity, metadataChanged = false, contentChanged = false)
    await(100.millis, withSystem = system)
    tracker.entityProcessed(entity)
    await(100.millis, withSystem = system)
    tracker.completed()

    updates.await.toList match {
      case first :: second :: third :: Nil =>
        first.entities.examined should be(Set(entity))
        first.entities.processed.keys.toSeq should be(empty)
        first.completed should be(empty)

        second.entities.examined should be(Set(entity))
        second.entities.processed.keys.toSeq should be(Seq(entity))
        second.completed should be(empty)

        third.entities.examined should be(Set(entity))
        third.entities.processed.keys.toSeq should be(Seq(entity))
        third.completed should not be empty

      case other =>
        fail(s"Unexpected result received: [$other]")
    }
  }

  it should "support dropping old state" in withRetry {
    val tracker = createTracker(maxRetention = 250.millis)

    val entity = Fixtures.Metadata.FileOneMetadata.path

    val operation1 = Operation.generateId()

    tracker.entityProcessed(entity)(operation1)
    tracker.completed()(operation1)

    eventually[Assertion] {
      tracker.state.await.keys.toSeq should be(Seq(operation1))
    }

    val operation2 = Operation.generateId()
    tracker.entityProcessed(entity)(operation2)

    eventually[Assertion] {
      tracker.state.await.keys.toSeq.sorted should be(Seq(operation1, operation2).sorted)
    }

    await(250.millis, withSystem = system)

    val operation3 = Operation.generateId()
    tracker.entityProcessed(entity)(operation3)

    eventually[Assertion] {
      tracker.state.await.keys.toSeq should be(Seq(operation3))
    }
  }

  it should "support providing at least one state update" in withRetry {
    val tracker = createTracker()

    implicit val operation: Operation.Id = Operation.generateId()

    val entity = Fixtures.Metadata.FileOneMetadata.path

    val initialState = tracker.state.await
    initialState should be(empty)

    val expectedUpdates = 1

    tracker.entityExamined(entity, metadataChanged = false, contentChanged = false)
    tracker.entityProcessed(entity)
    tracker.completed()
    await(100.millis, withSystem = system)

    val updates = tracker.updates(operation).take(expectedUpdates.toLong).runWith(Sink.seq)
    await(100.millis, withSystem = system)

    updates.await.toList match {
      case update :: Nil =>
        update.entities.examined should be(Set(entity))
        update.entities.processed.keys.toSeq should be(Seq(entity))
        update.completed should not be empty

      case other =>
        fail(s"Unexpected result received: [$other]")
    }
  }

  it should "support checking if operations exist" in {
    val tracker = createTracker()

    implicit val operation: Operation.Id = Operation.generateId()

    tracker.exists(operation).await should be(false)

    tracker.completed()

    eventually {
      tracker.exists(operation).await should be(true)
    }
  }

  it should "support removing operations" in withRetry {
    val tracker = createTracker(maxRetention = 250.millis)

    val entity = Fixtures.Metadata.FileOneMetadata.path

    val operation1 = Operation.generateId()
    val operation2 = Operation.generateId()

    tracker.entityProcessed(entity)(operation1)
    tracker.completed()(operation1)

    tracker.entityProcessed(entity)(operation2)

    eventually[Assertion] {
      tracker.state.await.keys.toSeq.sorted should be(Seq(operation1, operation2).sorted)
    }

    tracker.remove(operation1)

    eventually[Assertion] {
      tracker.state.await.keys.toSeq should be(Seq(operation2))
    }

    tracker.remove(operation2)

    eventually[Assertion] {
      tracker.state.await.keys.toSeq should be(empty)
    }
  }

  private def createTracker(maxRetention: FiniteDuration = 1.minute): DefaultRecoveryTracker = DefaultRecoveryTracker(
    maxRetention = maxRetention,
    createBackend = state =>
      EventLogMemoryBackend(
        name = s"test-recovery-tracker-${java.util.UUID.randomUUID()}",
        initialState = state
      )
  )(system.executionContext)

  override implicit val patienceConfig: PatienceConfig = PatienceConfig(7.seconds, 300.milliseconds)

  override implicit val timeout: Timeout = 7.seconds

  private implicit val system: ActorSystem[SpawnProtocol.Command] = ActorSystem(
    Behaviors.setup(_ => SpawnProtocol()): Behavior[SpawnProtocol.Command],
    "DefaultRecoveryTrackerSpec"
  )

  private implicit val telemetry: MockTelemetryContext = MockTelemetryContext()

  override protected def afterAll(): Unit =
    system.terminate()
}

File: client-android/app/src/androidTest/java/stasis/client_android/tracking/DefaultRecoveryTrackerSpec.kt
package stasis.client_android.tracking

import android.content.Context
import android.os.HandlerThread
import android.os.Process
import androidx.arch.core.executor.testing.InstantTaskExecutorRule
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import kotlinx.coroutines.runBlocking
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.CoreMatchers.not
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import stasis.client_android.Fixtures
import stasis.client_android.await
import stasis.client_android.eventually
import stasis.client_android.lib.model.TargetEntity
import stasis.client_android.lib.ops.Operation
import stasis.client_android.lib.ops.OperationId
import stasis.client_android.notNull
import java.nio.file.Paths
import java.util.UUID

@RunWith(AndroidJUnit4::class)
class DefaultRecoveryTrackerSpec {
    @Test
    fun trackRecoveryEvents() {
        val context = ApplicationProvider.getApplicationContext<Context>()

        val trackerHandler = HandlerThread(
            "DefaultRecoveryTrackerSpec",
            Process.THREAD_PRIORITY_BACKGROUND
        ).apply { start() }

        val tracker = DefaultRecoveryTracker(context, trackerHandler.looper)

        val operation: OperationId = UUID.randomUUID()
        val file1 = Paths.get("/tmp/test-1").toAbsolutePath()
        val file2 = Paths.get("/tmp/test-2").toAbsolutePath()

        val targetEntity = TargetEntity(
            path = file1,
            destination = TargetEntity.Destination.Default,
            existingMetadata = Fixtures.Metadata.FileOneMetadata,
            currentMetadata = null
        )

        val initialState = tracker.state.value

        assertThat(initialState, equalTo(emptyMap()))

        tracker.entityExamined(operation, entity = file1, metadataChanged = true, contentChanged = false)
        tracker.entityExamined(operation, entity = file2, metadataChanged = true, contentChanged = true)
        tracker.entityCollected(operation, entity = targetEntity)
        tracker.entityProcessingStarted(operation, entity = file1, expectedParts = 3)
        tracker.entityPartProcessed(operation, entity = file1)
        tracker.entityPartProcessed(operation, entity = file1)
        tracker.entityProcessed(operation, entity = file2)
        tracker.metadataApplied(operation, entity = file1)
        tracker.failureEncountered(operation, entity = file1, failure = RuntimeException("test failure 1"))
        tracker.failureEncountered(operation, failure = RuntimeException("test failure 2"))
        tracker.completed(operation)

        runBlocking {
            eventually {
                val completedState = tracker.state.await()[operation].notNull()

                assertThat(completedState.completed, not(equalTo(null)))

                assertThat(completedState.entities.examined.size, equalTo(2))
                assertThat(completedState.entities.collected.size, equalTo(1))
                assertThat(completedState.entities.pending.size, equalTo(1))
                assertThat(completedState.entities.processed.size, equalTo(1))
                assertThat(completedState.entities.metadataApplied.size, equalTo(1))
                assertThat(completedState.entities.failed.size, equalTo(1))

                assertThat(completedState.failures.size, equalTo(1))
            }
        }
    }

    @Test
    fun provideOperationUpdates() {
        val context = ApplicationProvider.getApplicationContext<Context>()

        val trackerHandler = HandlerThread(
            "DefaultRecoveryTrackerSpec",
            Process.THREAD_PRIORITY_BACKGROUND
        ).apply { start() }

        val tracker = DefaultRecoveryTracker(context, trackerHandler.looper)

        val operation: OperationId = UUID.randomUUID()
        val file = Paths.get("test").toAbsolutePath()

        val initialState = tracker.state.value

        assertThat(initialState, equalTo(emptyMap()))

        runBlocking {
            tracker.entityExamined(operation, entity = file, metadataChanged = true, contentChanged = false)

            eventually {
                val operationStateExamined = tracker.updates(operation).await()
                assertThat(operationStateExamined.entities.examined.size, equalTo(1))
                assertThat(operationStateExamined.completed, equalTo(null))
            }

            tracker.metadataApplied(operation = Operation.generateId(), entity = file) // other operation
            tracker.completed(operation)

            eventually {
                val operationStateCompleted = tracker.updates(operation).await()
                assertThat(operationStateCompleted.entities.examined.size, equalTo(1))
                assertThat(operationStateCompleted.completed, not(equalTo(null)))
            }
        }
    }

    @Test
    fun provideRecoveryState() {
        val context = ApplicationProvider.getApplicationContext<Context>()

        val trackerHandler = HandlerThread(
            "DefaultRecoveryTrackerSpec",
            Process.THREAD_PRIORITY_BACKGROUND
        ).apply { start() }

        val tracker = DefaultRecoveryTracker(context, trackerHandler.looper)

        val operation: OperationId = UUID.randomUUID()
        val file = Paths.get("test").toAbsolutePath()

        val initialState = tracker.state.value

        assertThat(initialState, equalTo(emptyMap()))


        runBlocking {
            tracker.entityExamined(operation, entity = file, metadataChanged = true, contentChanged = false)

            eventually {
                val state = tracker.stateOf(operation)
                assertThat(state?.entities?.examined?.size, equalTo(1))
                assertThat(state?.completed, equalTo(null))
            }
        }
    }

    @get:Rule
    val instantTaskExecutorRule = InstantTaskExecutorRule()
}
